bsip-0040.md

```diff
@@ -0,0 +1,322 @@
+
+### Custom active permission and custom active authority
+
+A `custom active permission` contains a list of `custom active authorities` and looks like follows (in JSON-like/pseudo for clarification):
+```
+custom_active_permission = {
+    account_id,    // account that is assigned to this permission
+    authorities = list of custom_active_authority objects
+}
+custom_active_authority = {
+    valid_from,    // timestamp when this is active, defaults to now
+    valid_to,      // timestamp when this is invalid, defaults to 1 year
+    operation_id,   // operation id of the target operation,
+    authority,     // same as for the existing authortities (multisig with weighted accounts or keys),
+    restrictions        // see below
+}
+```
+Note: This assumes `custom_active_permission` is stored in a separate index. Actual implementation details left to the implementer, as long as every `custom_active_permission` can be assigned to exactly one account.
+
+A `custom active authority` is matched against operations of an incoming, signed transaction. The wording *matching* refers to:
+- `operation_id` is equal to the id of the incoming operation
+- `account_id` is in the required accounts of the operation
+- the `authority` of the `custom_active_authority` is given by the signatures of the transaction
+- timestamp `now` is within `valid_to` and `valid_from` 
+- all `restrictions` assert positively
+
+### Restrictions
+
+The `restrictions` field is a list of restrictions consisting of argument to assert mappings.
+A dictionary-type object like
+```
+restriction = {
+   function,     // argument_identifier
+   argument,     // pointer to a dynamic value (argument of the operation, or attribute when nested)
+   data,         // data specific to the function
+}
+```
+is called a restriction, that can assert itself positively (passed) or negatively (violated). All restrictions are evaluated per default with `AND` logic to determine if the whole list has asserted positively. 
+
+List of possible restrictions are:
+
+| function        | data           | state  |
+| ------------- |:-------------:| -----:|
+| `any`      | [`list`, `of`, `allowed`, `values`] | stateless |
+| `none`     | [`none`, `of`, `these`, `values`] | stateless |
+| `lt, le, gt, ge`      | `comparative` | stateless |
+| `limit` | [`max_cumsum`, `interval_in_sec`]      | [`current_cumsum`, `interval_began`] |
+| `limit_monthly` | [`max_cumsum`, `interval_in_months`]      | [`current_cumsum`, `interval_began`] |
+| `attribute_assert` | list of restrictions | stateless |
+| `logical_or`     | list of restrictions lists | stateless |
+
+Following cases must hold for a restriction:
+- if there is no value given (e.g. an optional argument, or nested value not given), the restrictions is passed (no change, no violation)
+- if the expected type of the argument does not match the given type (no implicit type conversion), the restriction is violated
```

Comment:
Does it make sense to allow such restrictions at all?

Reply:
Is possible to check if the given restrictions match the argument types in the target operation?

With variable types in operation header this might not be possible and must be enforced on runtime, right?

Reply:
The argument types of restrictions should be check when installing the custom authority, at the validation step or even deserialization step. In an operation, data types are predefined, so they should match anyways.

Reply:
I have included that in Restrictions sections now